### PR TITLE
Fix AuthException when account contains no subscription

### LINKED ISSUE #2226: [IntelliJ][ReportedByUser] MS Services - Error Getting Subscriptions
IntelliJ build version: 2018.2.2 IU-182.4129.33
OS: Linux
JDK: JetBrains s.r.o 1.8.0_152-release
Plugin version: 3.13.0-2018.2
Error message: 
```
com.microsoft.azuretools.adauth.AuthException: No subscription found in the account
	at com.microsoft.azuretools.authmanage.SubscriptionManager.doSetSubscriptionDetails(SubscriptionManager.java:122)
	at com.microsoft.azuretools.authmanage.SubscriptionManager.updateSubscriptionDetailsIfNull(SubscriptionManager.java:92)
	at com.microsoft.azuretools.authmanage.SubscriptionManager.getSubscriptionDetails(SubscriptionManager.java:76)
	at com.microsoft.tooling.msservices.serviceexplorer.azure.AzureModule.composeName(AzureModule.java:105)
	at com.microsoft.tooling.msservices.serviceexplorer.azure.AzureModule.handleSubscriptionChange(AzureModule.java:220)
	at com.microsoft.tooling.msservices.serviceexplorer.azure.AzureModule.access$100(AzureModule.java:47)
	at com.microsoft.tooling.msservices.serviceexplorer.azure.AzureModule$SignInOutListener.run(AzureModule.java:229)
	at com.microsoft.azuretools.authmanage.AuthMethodManager.notifySignInEventListener(AuthMethodManager.java:97)
	at com.microsoft.azuretools.ijidea.actions.AzureSignInAction.doSignIn(AzureSignInAction.java:124)
	at com.microsoft.azuretools.ijidea.actions.AzureSignInAction.onAzureSignIn(AzureSignInAction.java:107)
	at com.microsoft.intellij.serviceexplorer.azure.SignInOutAction$1.actionPerformed(SignInOutAction.java:46)
	at com.microsoft.tooling.msservices.serviceexplorer.NodeActionListener.actionPerformedAsync(NodeActionListener.java:72)
	at com.microsoft.tooling.msservices.serviceexplorer.NodeAction.fireNodeActionEvent(NodeAction.java:60)
	at com.microsoft.intellij.components.ServerExplorerToolWindowFactory.lambd

### PR DESCRIPTION
**Pull request by Flanker32**
- Remove `AuthException`and handle related exception when there is no subscriptions in user account 
    * Now when users auth with an empty account, we will keep silent on auth and refresh, but show error prompt for action which need azure subscription
- Pop up info dialog when find no subscriptions after sign-in/select subscriptopm
- Fix Spring Cloud deployment panel refresh behavior

image

image

image

fixes #2226

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/container/webapponlinux/WebAppOnLinuxDeployConfiguration.java [hunk 1]
```diff
@@ -108,9 +104,7 @@ public RunProfileState getState(@NotNull Executor executor, @NotNull ExecutionEn
      */
     @Override
     public void validate() throws ConfigurationException {
-        if (!AuthMethodManager.getInstance().isSignedIn()) {
-            throw new ConfigurationException(NEED_SIGN_IN);
-        }
+        ValidationUtils.validateAuthentication();
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/container/webapponlinux/WebAppOnLinuxDeployConfiguration.java:107]: ensureLogin and choose another util class, not ValidationUtils

Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/container/webapponlinux/WebAppOnLinuxDeployConfiguration.java:107]: ensureLoginAndSubs

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/creation/FunctionCreationDialog.java [hunk 2]
```diff
@@ -174,6 +177,13 @@ public void init() {
         // Validate azure status
         if (!AuthMethodManager.getInstance().isSignedIn()) {
             res.add(new ValidationInfo("Please sign in with your Azure account.", subscriptionPanel.getComboComponent()));
+            return res;
+        }
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/deploy/ui/creation/FunctionCreationDialog.java:181]: should use an util method for login and non empty subscription

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/slimui/creation/WebAppCreationDialog.java [hunk 3]
```diff
@@ -297,6 +298,12 @@ protected void init() {
         // Validate azure status
         if (!AuthMethodManager.getInstance().isSignedIn()) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/slimui/creation/WebAppCreationDialog.java:299]: use util method

File: Utils/azuretools-core/src/com/microsoft/azuretools/core/mvp/model/AzureMvpModel.java [hunk 4]
```diff
@@ -125,6 +126,11 @@ public Subscription getSubscriptionById(String sid) {
         return ret;
     }
 
+    public boolean isSubscriptionSelected() {
+        final List<Subscription> subscriptionList = getSelectedSubscriptions();
```
Comment (andxu) [inline Utils/azuretools-core/src/com/microsoft/azuretools/core/mvp/model/AzureMvpModel.java:130]: no need for two line code, merge them.

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/AzureModule.java [hunk 5]
```diff
@@ -122,21 +120,20 @@ private static String composeName() {
...
                     .filter(SubscriptionDetail::isSelected).collect(Collectors.toList());
-            if (selectedSubscriptions.size() > 0) {
-                return String.format("%s (%s)", BASE_MODULE_NAME, selectedSubscriptions.size() > 1
-                        ? String.format("%s subscriptions", selectedSubscriptions.size())
-                        : selectedSubscriptions.get(0).getSubscriptionName());
+            final int subsCount = selectedSubscriptions.size();
+            switch (subsCount) {
+                case 0:
+                    return String.format("%s %s", BASE_MODULE_NAME, MODULE_NAME_NO_SUBSCRIPTION);
+                case 1:
```
Comment (andxu) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/AzureModule.java:127]: ```return String.format("%s (%s)", BASE_MODULE_NAME, getSubscriptionsDesc(selectedSubscriptions)```
and in `getSubscriptionsDesc` use switch statement.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/ValidationUtils.java [hunk 6]
```diff
@@ -127,6 +133,15 @@ public static void validateAppServicePlanName(String appServicePlan) {
         }
     }
 
+    public static void validateAuthentication() throws ConfigurationException {
+        if (!AuthMethodManager.getInstance().isSignedIn()) {
+            throw new ConfigurationException(NEED_SIGN_IN);
+        }
+        if (CollectionUtils.isEmpty(AzureMvpModel.getInstance().getSelectedSubscriptions())) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/ValidationUtils.java:140]: how can an user select EMPTY subs when he has some subs?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/ValidationUtils.java [hunk 7]
```diff
@@ -127,6 +133,15 @@ public static void validateAppServicePlanName(String appServicePlan) {
...
+            throw new ConfigurationException(CommonConst.MUST_SELECT_AN_AZURE_SUBSCRIPTION_FIRST);
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/ValidationUtils.java:141]: If the user has no subs, then error message is not applicable

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/ValidationUtils.java [hunk 8]
```diff
@@ -127,6 +133,15 @@ public static void validateAppServicePlanName(String appServicePlan) {
         }
     }
 
+    public static void validateAuthentication() throws ConfigurationException {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/ValidationUtils.java:136]: validateAuthentication is a bad name

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/slimui/creation/WebAppCreationDialog.java [hunk 9]
```diff
@@ -297,6 +298,12 @@ protected void init() {
...
             res.add(new ValidationInfo("Please sign in with your Azure account.", cbSubscription));
+            return res;
+        }
+        // Validate selected subscription count
+        if (AzureMvpModel.getInstance().isSubscriptionSelected()) {
+            res.add(new ValidationInfo("Please select subscription in azure explorer first.", cbSubscription));
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/webapp/webappconfig/slimui/creation/WebAppCreationDialog.java:305]: so many message "Please select subscription in ..."

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/AzureUtils.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+import com.microsoft.azure.common.exceptions.AzureExecutionException;
+import com.microsoft.azuretools.authmanage.AuthMethodManager;
+import com.microsoft.azuretools.authmanage.models.SubscriptionDetail;
+import com.microsoft.azuretools.sdkmanage.AzureManager;
+import org.apache.commons.collections.CollectionUtils;
+
+import java.io.IOException;
+import java.util.List;
+
+public class AzureUtils {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/AzureUtils.java:33]: AzureLoginHelper

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/AzureUtils.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+    public static final String NEED_SIGN_IN = "Please sign in with your Azure account.";
+    public static final String NO_SUBSCRIPTION = "No subscription in current account, you may get a free one from "
+            + "https://azure.microsoft.com/en-us/free/";
+    public static final String MUST_SELECT_SUBSCRIPTION =
+            "Please select at least one subscription first (Tools -> Azure -> Select Subscriptions)";
+
+    public static void checkAzurePreconditions() throws AzureExecutionException {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/AzureUtils.java:41]: ensureOneSubscriptionSelected

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/AzureUtils.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+import com.microsoft.azuretools.authmanage.models.SubscriptionDetail;
+import com.microsoft.azuretools.sdkmanage.AzureManager;
+import org.apache.commons.collections.CollectionUtils;
+
+import java.io.IOException;
+import java.util.List;
+
+public class AzureUtils {
+
+    public static final String NEED_SIGN_IN = "Please sign in with your Azure account.";
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/AzureUtils.java:35]: private

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/AzureUtils.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+                    azureManager.getSubscriptionManager().getSelectedSubscriptionDetails();
+            if (CollectionUtils.isEmpty(selectedSubscriptions)) {
+                throw new AzureExecutionException(MUST_SELECT_SUBSCRIPTION);
+            }
+        } catch (IOException e) {
+            throw new AzureExecutionException(e.getMessage(), e);
+        }
+    }
+
+    public static boolean checkAzurePreconditionsAndPrompt(String dialogTitle) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/util/AzureUtils.java:62]: ensureAzureSubsAvailableOrReportError

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/AzureModule.java [hunk 14]
```diff
@@ -122,21 +120,13 @@ private static String composeName() {
...
-            if (e instanceof AuthException &&
-                    StringUtils.equalsIgnoreCase(e.getMessage(), "No subscription found in the account")) {
-                return String.format("%s %s", BASE_MODULE_NAME, MODULE_NAME_NO_SUBSCRIPTION);
-            } else {
-                String msg = "An error occurred while getting the subscription list." + "\n" + "(Message from Azure:" + e
-                        .getMessage() + ")";
-                DefaultLoader.getUIHelper().showException(msg, e,
-                        "MS Services - Error Getting Subscriptions", false, true);
-            }
+            String msg = "An error occurred while getting the subscription list." + "\n" + "(Message from Azure:" + e
```
Comment (andxu) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/AzureModule.java:126]: better concatenation of string or format

File: Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/AzureModule.java [hunk 15]
```diff
@@ -276,6 +266,18 @@ public void run() {
         }
     }
 
+    private static String getAccountDescription(List<SubscriptionDetail> selectedSubscriptions) {
+        final int subsCount = selectedSubscriptions.size();
+        switch (subsCount) {
+            case 0:
+                return MODULE_NAME_NO_SUBSCRIPTION;
```
Comment (andxu) [inline Utils/azure-explorer-common/src/com/microsoft/tooling/msservices/serviceexplorer/azure/AzureModule.java:273]: remove brace in MODULE_NAME_NO_SUBSCRIPTION

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/ui/components/AzureDialogWrapper.java [hunk 16]
```diff
@@ -195,10 +213,27 @@ public SubscriptionDetail getSubscription() {
...
+                properties.put("SubscriptionId",
+                               this.getSubscription()
+                                   .getSubscriptionId());
+            }
         }
 
         return properties;
     }
+
+    protected ValidationInfo checkAzurePreconditions(JComponent component) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/ui/components/AzureDialogWrapper.java:231]: validateAzureSubsAvailable

Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/ui/components/AzureDialogWrapper.java]: Changed to `validateAzureSubs`

**Pull request merged**